Add Managing Child Objects guidance and examples (#3047)
grommet/hpe-design-system: aries-site/src/pages/templates/forms/managing-child-objects.mdx

|                 **1**                  | **Container**      | Identfies the child and controls the show detail behavior. | <Checkmark a11yTitle="true" size="small" /> | The entirety of the container is focusable and expands to show detail with a mouse click, Enter key, or Spacebar. The aria-label should read "Show and edit detail." |
|                 **2**                  | **Label**          | Uniquely identifies each child object.                     | <Checkmark a11yTitle="true" size="small" /> | Should be implemented as a semantically correct heading to allow screen reader users to traverse between children.                                                   |
|                 **3**                  | **Expand icon**    | Visual treatment indicating show detail behavior.          | <Checkmark a11yTitle="true" size="small" /> | Icon name is "FormDown".                                                                                                                                             |

> |                 **4**                  | **Values summary** | A preview summarizing the object's detail.                 |                  Optional                   |                                                                                                                                                                      |

Reviewer: for values summary should we include when to truncate? should they use grommet `truncation` or is there a recommended length?

Reviewer 2: Grommet truncation. I can clarify.

Reviewer: okay maybe we can add a small comment saying to should be implemented as grommets `text` with the `truncation` prop to allow for correct behavior. (something along those lines)

| **4** | **Body**          | Flexible region for presenting an object's detail.          | <Checkmark a11yTitle="true" size="small" /> | Detail presentation may vary by the use case.                                                                                                                                                                                                                                                             |
| **5** | **Remove button** | Removes/deletes the child from its parent.                  |               Case depending                | <span>Always provide unless at least one child object is required by the parent. See [Required children](#required-children) for detail.</span> Use the verb "remove" when removing the association between parent and child. Reserve the verb "delete" for actions where the child will no longer exist. |

## Examples

### Optional children

Use cases where child objects are not required by the parent may use a semantically correct
heading, optional description, and a secondary button labelled with an add action to the form.

<Example
  code={[

>     'https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/forms/managing-child-objects/CreateRole.js',

Reviewer 2: Done.

import { Example } from '../../../layouts';

> ## When to use

Reviewer: This is great content and very informal however some of this to me sounds more `why to use` rather then `when to use`  

for example:
This pattern improves productivity by focusing a user's attention solely on the current object of interest, and uses collapsible objects to show or hide detail as needed. The intent is that collapsible objects reduce noise, clutter, and the resulting cognitive overload by hiding extraneous detail — and an easy means to view detail when needed.

I would reduce this with:
Use this pattern when a user needs to expand/collapse on different child objects as an easy means to view detail when needed. (can work on wording) 

```

This pattern improves productivity by focusing a user's attention solely on the current object of interest, and uses collapsible objects to show or hide detail as needed. The intent is that collapsible objects reduce noise, clutter, and the resulting cognitive overload by hiding extraneous detail 

```

This is more informed about the component pattern itself rather than `when to use`
just my opinion, unless we want this in this section just thinking out loud

Author: Good comments. I tend to agree. I originally had they "why" portion above as part of the purpose statement, but it didn't flow as well. I don't want to delete it outright either, as I think we need our content to educate readers as to the rationale/why so that they have the deeper level of understanding.

I'll take another pass.

Reviewer: yeah totally agree it is great information and I wouldn't want to take it out. I also thought about a `why` section but its hard since some of our pages don't have similar flow I was thinking in terms of how the `layer` page is getting done but did not see a good section in there. Happy to brainstorm as well if depending on second pass

Author: @britt6612 - I reworked this a bit. See if it flows better for you now.

The managing child objects pattern may be used when creating, updating, or removing 
child objects is part of a larger flow where a parent resource is being created or 

> configured. In these circumstances, the additional configuration inputs can crowd screens 

Reviewer 2: ```suggestion
configured. In these circumstances, displaying all of the additional configuration inputs at once can crowd screens 
```

Author: done

and lead to quite lengthy forms.

This pattern improves productivity by focusing a user's attention solely on the current object 
of interest, and uses collapsible objects to show or hide detail as needed. The 

> intent is that collapsible objects reduce noise, clutter, and the resulting cognitive overload 

Reviewer 2: ```suggestion
intent is that collapsible objects reduce noise, clutter, and cognitive overload 
```
nit, but felt like could be simplified to this

Author: Like. 👍

Author: done

by hiding extraneous detail &mdash; and an easy means to view detail when needed.

## Anatomy

The following illustrates the elements composing a child object in its collapsed and expanded 
states. View the [use case examples](#use-case-examples) to see how child objects behave and are presented in a form 
context.

### Collapsed state

[Diagram to be inserted]

> | Label | Region         | Purpose                                                    | Required? | Notes |

Reviewer 3: ```suggestion
| Label | Region         | Purpose                                                    | Required | Notes |
```
It feels a little strange to have the question mark in the column header

Author: Agree. I also don't really like the repetitive "required." I was aligning with what has been done on the Layer work, but what if we tried something like:
- column header reads "Required"
- values read either "true" or "<Checkmark a11yTitle="true" />," "optional," or blank
?

cc: @taysea

Reviewer 2: I'm open to that as well, I think I'd tried the true/false approach locally in layers but then found myself scanning back up to the column header to see what "true" / "false" was referring too.

Totally fine giving that a go though if it feels less repetitive.

Author: done

| ----- | -------------- | ---------------------------------------------------------- | --------- | ----- |

> |       | Container      | Identfies the child and controls the show detail behavior. | Required  | The entirety of the container is focusable and shows detail with a mouse click, Enter key, or Spacebar. The aria-label should read "Show and edit detail." |

Reviewer 2: ```suggestion
|       | Container      | Identfies the child and controls the show detail behavior. | Required  | The entirety of the container is focusable and expands to show detail with a mouse click, Enter key, or Spacebar. The aria-label should read "Show and edit detail." |
```

Author: 👍

Author: done

[Diagram to be inserted]

| Label | Region        | Purpose                                                     | Required? | Notes |
| ----- | ------------- | ----------------------------------------------------------- | --------- | ----- |
|       | Header        | Identifies the child and controls the hide detail behavior. | Required  | The entirety of the container is focusable and hides detail with a mouse click, Enter key, or Spacebar. The aria-label should read "Hide detail." |
|       | Label         | Uniquely identifies each child object.                      | Required  | Should be implemented as a semantically correct heading to allow screen reader users to traverse between children. |
|       | Collapse icon | Visual treatment indicating hide detail behavior.           | Required  | Icon name is "FormUp". |
|       | Body          | Flexible region for presenting an object's detail.          | Required  | Detail presentation may vary by the use case. |
|       | Remove button | Removes/deletes the child from its parent.                  | Required  |       | 
|       | Border        | Provides visual distinction from other children.            | Required  |       |

> ## Use case examples

Reviewer 2: Small alignment thing -- in cards and layer I've just been calling this section "Examples". Thinking we should align across pages. Do you have a strong preference in one direction or the other? Or do you view Examples vs Use case examples as distinct things?

Author: 🤔  -- I noticed that when reviewing the layer work as well. 

I'm not sure which way to think about it. "Examples" provides more flexibility as it would be inclusive of "example use cases" but allow for non-use case oriented executions (having a hard time of thinking what those might be at the moment).

Perhaps we simplify with "Examples" for the time being?

Reviewer 2: I'm good with that.

Author: Done.

</Example>

| Label | Region         | Purpose                                                    | Required | Notes |
| ----- | -------------- | ---------------------------------------------------------- | :------: | ----- |
|       | Container      | Identfies the child and controls the show detail behavior. | <Checkmark a11yTitle="true" size="small" /> | The entirety of the container is focusable and expands to show detail with a mouse click, Enter key, or Spacebar. The aria-label should read "Show and edit detail." |
|       | Label          | Uniquely identifies each child object.                     | <Checkmark a11yTitle="true" size="small" /> | Should be implemented as a semantically correct heading to allow screen reader users to traverse between children. |
|       | Expand icon    | Visual treatment indicating show detail behavior.          | <Checkmark a11yTitle="true" size="small" /> | Icon name is "FormDown". | 
|       | Values summary | A preview summarizing the object's detail.                 | Optional |       |
|       | Border         | Provides visual distinction from other children.           | <Checkmark a11yTitle="true" size="small" /> |       |

> ### Expanded state

Reviewer 4: Would we want to include details on keyboard interactions within this section? Specifically info on how the container collapses once the user tabs past the remove button.

|       | Body          | Flexible region for presenting an object's detail.          | <Checkmark a11yTitle="true" size="small" /> | Detail presentation may vary by the use case. |
|       | Remove button | Removes/deletes the child from its parent.                  | <Checkmark a11yTitle="true" size="small" /> | Use the verb "remove" when removing the association between parent and child. Reserve the verb "delete" for actions where the child will no longer exist. | 
|       | Border        | Provides visual distinction from other children.            | <Checkmark a11yTitle="true" size="small" /> |       |

## Examples

### Optional children

Use cases where child objects are not required by the parent may use a semantically correct 
heading, optional description, and a secondary button labelled with an add action to the form.

[Example to be inserted]

### Requried children

For convenience, use cases where a parent can not be created without associating at least one 
child (e.g. creating a cluster) may present the first child object in its expanded state.

> <Example 

Author: I have that as the button's `aria-label` but will add it to the visual label as well.

> <Example 

Author: Done.
